### ISSUE TITLE
.maintenance.flag and .maintenance.ip files are stored in, and queried from S3 bucket, when S3 storage is enabled.

### ISSUE THREAD

**Issue by gwharton** [magento/magento2#39047]
### Preconditions and environment

2.4.7-p1
S3 Storage Enabled

### Steps to reproduce

Deploy 2.4.7-p1
Enable S3 Storage
Enable Maintenance Mode

### Expected result

I wouldn't expect the mechanism to determine whether Magento is in maintenance mode or not to require a remote call to the S3 bucket.

### Actual result

Every time my server is hit at the frontend, it fires off an API call to the S3 bucket to check for .maintenance.flag

### Additional information

Not a bug as such, but feels like something that crept in without noticing when remote storage was envisaged.

I'm sure there are use cases to have a central .maintenance.flag shared across all instances

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by makkoff** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Hotel** [comment]
Hello @gwharton,

Thanks for the report and collaboration!

It seems to us an expected behavior to check whether the Magento is in maintenance mode or not. And because we enabled the S3, all assets are kept in the S3 bucket only that's why an API has been triggering to check the mode of Magento.

Please let us know if we have missed anything here.

Thanks

**Comment by gwharton** [comment]
I agree it is the exected behaviour given the current implementation. 

I was however querying whether it is the *right* behaviour.

API calls to S3 are chargable per call.

Several calls are made to S3 to check if maintenance mode is enabled for every instantiation of the Magento App which both increases costs, and also increases latency to the end user.

**Comment by engcom-Hotel** [comment]
Hello @gwharton,

I agree with your point that S3 API calls have been chargeable. Keeping this in mind, I am marking this issue as a feature request for further processing.

Thanks

**Comment by gwharton** [comment]
Yeah, not a bug as such, just could do with improvement.

The functionality is well abstracted, so would seem possible to provide a configurable maintenance flag system with options to use

1. Automatic Filesystem (local or remote if enabled, current implementation)
2. Flag table in database (supports multi node, is the DB configured this early in the bootstrap?)
3. Local Filesystem (single node)
4. Remote Filesystem (if remote enabled, supports multi node)

Perhaps configurable through di.xml with the default being the current implementation.
